### PR TITLE
Migrate same package name

### PR DESCRIPTION
**Pull request by robin-da**
Some improvements to make the migration process run when migrating
between packages with the same name, e.g. foo-1.0.0 -> foo-2.0.0. This
is the main use case.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-assistant/integration-tests/src/Main.hs [hunk 1]
```diff
@@ -278,31 +279,31 @@ packagingTests tmpDir = testGroup "packaging"
...
-        assertBool "b.dar was not created." =<< doesFileExist bDar
-        step "Creating migration project"
-        callCommandQuiet $ unwords ["daml", "migrate", projectMigrate, "daml/Main.daml", aDar, bDar]
+        withCurrentDirectory projectB $ callCommandQuiet "daml build"
+        assertBool "a-2.0.dar was not created." =<< doesFileExist bDar
+        step "Creating upgrade project"
+        callCommandQuiet $ unwords ["daml", "migrate", projectUpgrade, "daml/Main.daml", aDar, bDar]
         step "Build migration project"
-        withCurrentDirectory projectMigrate $ callCommandQuiet $ "daml build --output " <> distDir </> "c.dar"
+        withCurrentDirectory projectUpgrade $ callCommandQuiet "daml build --init-package-db=no --package \'(\"a-1.0\", [(\"Main\", \"MainA\")])\' --package \'(\"a-2.0\", [(\"Main\", \"MainB\")])\'"
```
Comment (cocreature) [inline daml-assistant/integration-tests/src/Main.hs]: Why do you need `--init-package-db=no`? Is this an optimization because `daml migrate` already takes care off this or a requirement because running `daml init` again will mess things up in some way?

File: daml-assistant/integration-tests/src/Main.hs [hunk 2]
```diff
@@ -231,11 +231,12 @@ packagingTests tmpDir = testGroup "packaging"
     , testCaseSteps "Build migration package" $ \step -> do
         let projectA = tmpDir </> "a"
         let projectB = tmpDir </> "b"
-        let projectMigrate = tmpDir </> "migrateAB"
-        let aDar = projectA </> distDir </> "a.dar"
-        let bDar = projectB </> distDir </> "b.dar"
-        let bUpgradedDar = tmpDir </> "b_upgraded.dar"
-        step "Creating project a..."
+        let projectUpgrade = tmpDir </> "upgrade"
+        let aDar = projectA </> distDir </> "a-1.0.dar"
```
Comment (cocreature) [inline daml-assistant/integration-tests/src/Main.hs]: What happens if I pass in two dars with the same version but different contents? Do I get a sensible error message or do things blow up badly? Might also makes sense to add a test for that.

File: daml-assistant/integration-tests/src/Main.hs [hunk 3]
```diff
@@ -278,31 +279,31 @@ packagingTests tmpDir = testGroup "packaging"
...
-        assertBool "b_upgraded.dar was not created." =<< doesFileExist bUpgradedDar
+        callCommandQuiet $
+          unwords
+              [ "daml damlc merge-dars"
+              , bDar
+              , upgradeDar
+              , "--dar-name"
+              , bWithUpgradesDar
+              ]
+        assertBool "a-0.2-with-upgrades.dar was not created." =<< doesFileExist bWithUpgradesDar
```
Comment (cocreature) [inline daml-assistant/integration-tests/src/Main.hs]: We really need to add a test that we can actually deploy this dar and run the upgrade but let’s keep that for a separate PR.

File: daml-assistant/integration-tests/src/Main.hs (hunk 1 above)
Comment (robin-da) [reply daml-assistant/integration-tests/src/Main.hs]: yes, `migrate` takes care of setting up a sensible package database. `daml init` would mess things up.

File: daml-assistant/integration-tests/src/Main.hs (hunk 2 above)
Comment (robin-da) [reply daml-assistant/integration-tests/src/Main.hs]: Compilation of the migration code would fail, package names need to be different, otherwise you will get overlapping instances.

File: daml-assistant/integration-tests/src/Main.hs (hunk 3 above)
Comment (robin-da) [reply daml-assistant/integration-tests/src/Main.hs]: agreed!

File: daml-assistant/integration-tests/src/Main.hs (hunk 1 above)
Comment (cocreature) [reply daml-assistant/integration-tests/src/Main.hs]: In that case, should we turn `migrate + build` into a single command (with a --no-build option for debugging)? It seems a bit unintuitive that`daml build` doesn’t work without special flags.

Comment (robin-da) [reply daml-assistant/integration-tests/src/Main.hs]: I now changed it, so that the migration project contains a `build.sh` that you can run. Generating the migration project and building it are two different processes. That the build will fail, will be a common thing, since the newer package might contain non-isomorphic contract templates. So I would not couple the two commands and instead, provide an easy way to try the build again once changes have been made manually to the source.

Comment (cocreature) [reply daml-assistant/integration-tests/src/Main.hs]: Sounds good, thanks!

**Comment by cocreature** [comment]
@robin-da It looks like the escaping now works (at least as long as you don’t start to include double or single quotes in package or version names. Not entirely sold on the syntax of `--package` but as long as we don’t ask users to type this it’s probably fine for now.

**Comment by robin-da** [comment]
yes, we should definitely improve the package syntax. I think we might be able to reuse ghc's flags and syntax and drop the flag entirely.
